File: client/baron.go
package main

import (
	"math"
	"math/rand"
	"strconv"

	"../fast"

	"./render"
)

const (
	baronSleep   = uint8(0)
	baronDead    = uint8(1)
	baronLook    = uint8(2)
	baronChase   = uint8(3)
	baronMelee   = uint8(4)
	baronMissile = uint8(5)
)

var (
	baronAnimationIdle    [][]*render.Sprite
	baronAnimationWalk    [][]*render.Sprite
	baronAnimationMelee   [][]*render.Sprite
	baronAnimationMissile [][]*render.Sprite
	baronAnimationDeath   [][]*render.Sprite
)

type baron struct {
	*thing
	status uint8
}

func baronInit(world *world, nid uint16, x, y, z float32, direction uint8, health uint16, status uint8) *baron {
	baron := &baron{}
	baron.thing = &thing{}
	baron.thing.update = baron.updateFn
	baron.world = world
	baron.uid = BaronUID
	baron.sid = "baron"
	baron.nid = nid
	baron.animateInit(status)
	baron.x = x
	baron.y = y
	baron.z = z
	if direction != DirectionNone {
		baron.angle = DirectionToAngle[direction]
	}
	baron.oldX = x
	baron.oldY = y
	baron.oldZ = z
	baron.radius = 0.4
	baron.height = 1.0
	baron.speed = 0.1
	baron.health = health
	baron.status = status
	world.addThing(baron.thing)
	world.netLookup[baron.nid] = baron
	baron.blockBorders()
	baron.addToBlocks()
	return baron
}

func (me *baron) animateInit(status uint8) {
	switch status {
	case baronDead:
		me.animation = baronAnimationDeath
	case baronMelee:
		me.animation = baronAnimationMelee
	case baronMissile:
		me.animation = baronAnimationMissile
	default:
		me.animation = baronAnimationWalk
	}
}

func (me *baron) netUpdate(data *fast.ByteReader, delta uint8) {
	me.thingNetUpdate(data, delta)
	if delta&0x4 != 0 {
		me.netUpdateHealth(data.GetUint16())
	}
	if delta&0x8 != 0 {
		me.netUpdateState(data.GetUint8())
	}
	if delta&0x10 != 0 {
		direction := data.GetUint8()
		if direction != DirectionNone {
			me.angle = DirectionToAngle[direction]
		}
	}
}

func (me *baron) netUpdateState(status uint8) {
	if me.status == status {
		return
	}
	me.animationMod = 0
	me.animationFrame = 0
	switch status {
	case baronDead:
		me.animation = baronAnimationDeath
	case baronMelee:
		me.animation = baronAnimationMelee
		playWadSound("baron-melee")
	case baronMissile:
		me.animation = baronAnimationMissile
		playWadSound("baron-missile")
	case baronChase:
		if rand.Float32() < 0.1 {
			playWadSound("baron-scream")
		}
	default:
		me.animation = baronAnimationWalk
	}
	me.status = status
}

func (me *baron) netUpdateHealth(health uint16) {
	if health < me.health {
		if health < 1 {
			playWadSound("baron-death")
		} else {
			playWadSound("baron-pain")
		}
		for i := 0; i < 20; i++ {
			spriteName := "blood-" + strconv.Itoa(int(math.Floor(rand.Float64()*3)))
			x := me.x + me.radius*(1-rand.Float32()*2)
			y := me.y + me.height*rand.Float32()
			z := me.z + me.radius*(1-rand.Float32()*2)
			const spread = 0.2
			dx := spread * (1 - rand.Float32()*2)
			dy := spread * rand.Float32()
			dz := spread * (1 - rand.Float32()*2)
			bloodInit(me.world, x, y, z, dx, dy, dz, spriteName)
		}

	}
	me.health = health
}

func (me *baron) dead() {
	if me.animationFrame == len(me.animation)-1 {
		me.update = me.emptyUpdate
	} else {
		me.updateAnimation()
	}
}

func (me *baron) look() {
	if me.updateAnimation() == AnimationDone {
		me.animationFrame = 0
	}
}

func (me *baron) melee() {
	if me.updateAnimation() == AnimationDone {
		me.animationFrame = 0
		me.animation = baronAnimationWalk
	}
}

func (me *baron) missile() {
	if me.updateAnimation() == AnimationDone {
		me.animationFrame = 0
		me.animation = baronAnimationWalk
	}
}

func (me *baron) chase() {
	if me.updateAnimation() == AnimationDone {
		me.animationFrame = 0
	}
}

func (me *baron) updateFn() {
	switch me.status {
	case baronDead:
		me.dead()
	case baronLook:
		me.look()
	case baronMelee:
		me.melee()
	case baronMissile:
		me.missile()
	case baronChase:
		me.chase()
	}
	me.updateNetworkDelta()
}

func (me *baron) emptyUpdate() {
}

File: public/baron.js
const BaronAnimationIdle = []
const BaronAnimationWalk = []
const BaronAnimationMelee = []
const BaronAnimationMissile = []
const BaronAnimationDeath = []

const BaronSleep = 0
const BaronDead = 1
const BaronLook = 2
const BaronChase = 3
const BaronMelee = 4
const BaronMissile = 5

class Baron extends Thing {
    constructor(world, nid, x, y, z, direction, heatlh, status) {
        super()
        this.World = world
        this.UID = BaronUID
        this.SID = "baron"
        this.NID = nid
        this.Update = this.BaronUpdate
        this.Animation = this.GetAnimation(status)
        this.X = x
        this.Y = y
        this.Z = z
        this.Angle = DirectionToAngle[direction]
        this.OldX = x
        this.OldY = y
        this.OldZ = z
        this.Radius = 0.4
        this.Height = 1.0
        this.Speed = 0.1
        this.Health = heatlh
        this.Status = status
        world.AddThing(this)
        this.BlockBorders()
        this.AddToBlocks()
    }
    Save() {
        let data = "{u:" + this.UID
        data += ",x:" + this.X
        data += ",y:" + this.Y
        data += ",z:" + this.Z
        data += ",a:" + this.Angle
        data += ",h:" + this.Health
        data += "}"
        return data
    }
    GetAnimation(status) {
        switch (status) {
            case BaronDead:
                return BaronAnimationDeath
            case BaronMelee:
                return BaronAnimationMelee
            case BaronMissile:
                return BaronAnimationMissile
            default:
                return BaronAnimationWalk
        }
    }
    NetUpdateState(status) {
        if (this.Status === status)
            return
        this.AnimationMod = 0
        this.AnimationFrame = 0
        switch (status) {
            case BaronDead:
                this.Animation = BaronAnimationDeath
                break
            case BaronMelee:
                this.Animation = BaronAnimationMelee
                PlaySound("baron-melee")
                break
            case BaronMissile:
                this.Animation = BaronAnimationMissile
                PlaySound("baron-missile")
                break
            case BaronChase:
                if (Math.random() < 0.1)
                    PlaySound("baron-scream")
            default:
                this.Animation = BaronAnimationWalk
                break
        }
        this.Status = status
    }
    NetUpdateHealth(health) {
        if (health < this.Health) {
            if (health < 1) {
                PlaySound("baron-death")
            } else {
                PlaySound("baron-pain")
            }
            for (let i = 0; i < 20; i++) {
                let spriteName = "blood-" + Math.floor(Math.random() * 3)
                let x = this.X + this.Radius * (1 - Math.random() * 2)
                let y = this.Y + this.Height * Math.random()
                let z = this.Z + this.Radius * (1 - Math.random() * 2)
                const spread = 0.2
                let dx = spread * (1 - Math.random() * 2)
                let dy = spread * Math.random()
                let dz = spread * (1 - Math.random() * 2)
                new Blood(this.World, x, y, z, dx, dy, dz, spriteName)
            }

        }
        this.Health = health
    }
    Dead() {
        if (this.AnimationFrame === this.Animation.length - 1) {
            this.Update = this.EmptyUpdate
        } else {
            this.UpdateAnimation()
        }
    }
    Look() {
        if (this.UpdateAnimation() === AnimationDone) {
            this.AnimationFrame = 0
        }
    }
    Melee() {
        if (this.UpdateAnimation() === AnimationDone) {
            this.AnimationFrame = 0
            this.Animation = BaronAnimationWalk
        }
    }
    Missile() {
        if (this.UpdateAnimation() === AnimationDone) {
            this.AnimationFrame = 0
            this.Animation = BaronAnimationWalk
        }
    }
    Chase() {
        if (this.UpdateAnimation() === AnimationDone) {
            this.AnimationFrame = 0
        }
    }
    BaronUpdate() {
        switch (this.Status) {
            case BaronDead:
                this.Dead()
                break
            case BaronLook:
                this.Look()
                break
            case BaronMelee:
                this.Melee()
                break
            case BaronMissile:
                this.Missile()
                break
            case BaronChase:
                this.Chase()
                break
        }
        this.LerpNetCode()
    }
    EmptyUpdate() {}
}
